daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/types/Ledger.scala

```diff
@@ -552,36 +553,40 @@ object Ledger {
         )
     }
 
-    def divulgeContracts(witnesses: Set[Party], coids: Set[ContractId]): EnrichState =
-      coids.foldLeft(this) {
-        case (s, coid) => s.divulgeCoidTo(witnesses, coid)
-      }
+    def divulgeContracts(
+        localContracts: Map[TContractId, Transaction.NodeId],
+        witnesses: Set[Party],
+        coids: Set[ContractId],
+    ): EnrichState =
+      coids.foldLeft(this)(_.divulgeContract(localContracts, witnesses, _))
 
-    def divulgeCoidTo(witnesses: Set[Party], coid: ContractId): EnrichState = {
-      def divulgeRelativeCoidTo(ws: Set[Party], rcoid: RelativeContractId): EnrichState = {
-        val nid = rcoid.txnid
-        copy(
-          localDivulgences = localDivulgences
-            .updated(nid, ws union localDivulgences.getOrElse(nid, Set.empty)),
+    def divulgeContract(
+        localContracts: Map[TContractId, Transaction.NodeId],
+        witnesses: Set[Party],
+        coid: ContractId,
+    ): EnrichState =
+      localContracts
+        .get(coid)
+        .fold(
+          coid match {
+            case acoid: AbsoluteContractId =>
+              copy(
+                globalDivulgences = globalDivulgences
+                  .updated(acoid, witnesses union globalDivulgences.getOrElse(acoid, Set.empty))
+              )
+            case rcoid: RelativeContractId =>
+              throw new IllegalArgumentException(
+                s"relative contract id $rcoid does not match any known local contract")
```

Comment:
Would it be possible to take a `Map[AbsoluteContractId, Transaction.NodeId]` instead and forego this runtime check? If not, I would suggest at least moving this check before the `get`.

Reply:
We plan to drop `RelativeContractIds`. So from now on, the logic of new/patched code should not use `RelativeContractId` and relies only on the method `localContracts` from `GenTransaction`.

I consider this runtime check is not part of the logic, since it should never be triggered (if the transaction is well-formed) and is just here to adapt the type of  field `globalDivulgences`.
Moving the check before the get, move it inside the logic.

Reply:
There is obviously something I did not understand about the logic, though.